File: src/javascript/algorithms/hashmap/continousSubarraySum/continousSubarraySum.js
/**
 * Brute force TC: O(n²) SC: O(1)
 */
function checkSubarraySum1(nums, k) {
    for (let i = 0; i < nums.length - 1; i++) {
        let sum = nums[i];
        for (let j = i + 1; j < nums.length; j++) {
            sum += nums[j];
            if (sum % k === 0) return true;
        }
    }
    return false;
}

/**
 * Prefix sum + HashMap TC: O(n) SC: O(min(n, k))
 */
function checkSubarraySum2(nums, k) {
    const remainderMap = new Map([[0, -1]]);
    let total = 0;
    for (let i = 0; i < nums.length; i++) {
        total += nums[i];
        const remainder = total % k;
        if (!remainderMap.has(remainder)) {
            remainderMap.set(remainder, i);
        } else if (i - remainderMap.get(remainder) > 1) {
            return true;
        }
    }
    return false;
}

const testCases = [
    { nums: [23, 2, 4, 6, 7], k: 6,  expected: true },
    { nums: [23, 2, 6, 4, 7], k: 6,  expected: true },
    { nums: [23, 2, 6, 4, 7], k: 13, expected: false },
    { nums: [5, 0, 0, 0],     k: 3,  expected: true },
    { nums: [1, 0],           k: 2,  expected: false },
    { nums: [0, 0],           k: 1,  expected: true },
];

function runTests(fn, label) {
    console.log(`--- ${label} ---`);
    for (const { nums, k, expected } of testCases) {
        const result = fn([...nums], k);
        console.log(`[${result === expected ? "PASS" : "FAIL"}] result=${result} expected=${expected}`);
    }
}

runTests(checkSubarraySum1, "checkSubarraySum1 (brute force)");
runTests(checkSubarraySum2, "checkSubarraySum2 (prefix sum + hashmap)");

File: src/python/algorithms/hashmap/continousSubarraySum/continous_subarray_sum.py
from typing import List


def check_subarray_sum1(nums: List[int], k: int) -> bool:
    """Brute force TC: O(n²) SC: O(1)"""
    for i in range(len(nums) - 1):
        total = nums[i]
        for j in range(i + 1, len(nums)):
            total += nums[j]
            if total % k == 0:
                return True
    return False


def check_subarray_sum2(nums: List[int], k: int) -> bool:
    """Prefix sum + hashmap TC: O(n) SC: O(min(n, k))"""
    remainder_map = {0: -1}
    total = 0
    for i, num in enumerate(nums):
        total += num
        remainder = total % k
        if remainder not in remainder_map:
            remainder_map[remainder] = i
        elif i - remainder_map[remainder] > 1:
            return True
    return False


if __name__ == "__main__":
    test_cases = [
        ([23, 2, 4, 6, 7], 6,  True),
        ([23, 2, 6, 4, 7], 6,  True),
        ([23, 2, 6, 4, 7], 13, False),
        ([5, 0, 0, 0],     3,  True),
        ([1, 0],           2,  False),
        ([0, 0],           1,  True),
    ]

    for fn, label in [
        (check_subarray_sum1, "check_subarray_sum1 (brute force)"),
        (check_subarray_sum2, "check_subarray_sum2 (prefix sum + hashmap)"),
    ]:
        print(f"--- {label} ---")
        for nums, k, expected in test_cases:
            result = fn(nums[:], k)
            status = "PASS" if result == expected else "FAIL"
            print(f"[{status}] result={result} expected={expected}")
